File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {

        List<List<String>> ans = new ArrayList<>();
        List<String> path = new ArrayList<>();

        helper2(s, 0, path, ans);
        return ans;

    }

    void helper2(String s, int start, List<String> path, List<List<String>> ans) {

        if (start == s.length()) {
            ans.add(new ArrayList<>(path));
            return;
        }

        for (int i = start + 1; i <= s.length(); i++) {
            String curstring = s.substring(start, i);
            if (isPalindrome(s, start, i - 1)) {
                path.add(curstring);
                helper2(s, i, path, ans);
                path.remove(path.size() - 1);
            }
        }

    }

    void helper(String s, int index, List<String> path, List<List<String>> ans) {

        if (index == s.length()) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int i = index; i < s.length(); i++) {
            if (isPalindrome(s, index, i)) {
                path.add(s.substring(index, i + 1));
                helper(s, i + 1, path, ans);
                path.remove(path.size() - 1);
            }
        }
    }

    boolean isPalindrome(String s, int start, int end) {
        while (start <= end) {
            if (s.charAt(start++) != s.charAt(end--))
                return false;
        }

        return true;
    }
}

File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        
        cur = []
        ans = []
        self.helper(0, s, cur, ans)
        return ans

    def helper(self, ind, s, cur, ans):

        if ind == len(s):
            ans.append(cur.copy())
            return

        for i in range(ind, len(s)):
            sub_str = s[ind:i+1]
            if self.is_palindrome(sub_str):
                cur.append(sub_str)
                self.helper(i+1, s, cur, ans)
                cur.pop()
        
    def is_palindrome(self, s):
        return s == s[::-1]
        